libs/dh/admin/feature-create-user-role-modal/src/lib/dh-create-user-role-modal.component.ts

```diff
@@ -0,0 +1,223 @@
+
+  @Output() closed = new EventEmitter<{ saveSuccess: boolean }>();
+
+  permissions$ = this.marketRolePermissionsStore.permissions$;
+
+  isSubmitted = false;
+
+  userRoleForm = this.formBuilder.nonNullable.group<UserRoleForm>({
+    eicFunction: this.formBuilder.nonNullable.control(
+      MarketParticipantEicFunction.BalanceResponsibleParty,
+      Validators.required
+    ),
+    name: this.formBuilder.nonNullable.control('', [
+      Validators.required,
+      Validators.maxLength(250),
+    ]),
+    description: this.formBuilder.nonNullable.control('', Validators.required),
+    status: this.formBuilder.nonNullable.control(MarketParticipantUserRoleStatus.Active),
+  });
+
+  selectedPermissions = new FormControl<number[]>([], {
+    validators: Validators.required,
+    nonNullable: true,
+  });
+
+  eicFunctionOptions: WattDropdownOptions = [];
+
+  ngOnInit(): void {
+    this.buildEicFunctionOptions();
+
+    this.userRoleForm.controls.eicFunction.valueChanges
+      .pipe(takeUntil(this.destroy$))
+      .subscribe((value) => {
+        this.marketRolePermissionsStore.getPermissions(value);
+      });
+
+    this.marketRolePermissionsStore.getPermissions(this.userRoleForm.controls.eicFunction.value);
+  }
+
+  ngAfterViewInit(): void {
+    this.createUserRoleModal.open();
+  }
+
+  ngOnDestroy(): void {
+    this.destroy$.next();
+    this.destroy$.complete();
+  }
+
+  onSelectionChange(event: MarketParticipantPermissionDetailsDto[]): void {
+    const ids = event.map(({ id }) => id);
+
+    this.selectedPermissions.setValue(ids);
+    this.selectedPermissions.markAsTouched();
+  }
+
+  closeModal(saveSuccess: boolean) {
+    this.createUserRoleModal.close(saveSuccess);
+    this.closed.emit({ saveSuccess });
+  }
+
+  createUserRole(): void {
+    this.isSubmitted = true;
+
+    if (this.selectedPermissions.invalid) {
+      return;
+    }
+
+    const createUserRoleDto: MarketParticipantCreateUserRoleDto = {
+      ...this.userRoleForm.getRawValue(),
+      permissions: this.selectedPermissions.value,
+    };
+
+    this.createUserRoleStore.createUserRole({
+      createUserRoleDto,
+      onSuccessFn: this.onSuccesFn,
+      onErrorFn: this.onErrorFn,
+    });
+
+    this.toastService.open({
+      message: this.transloco.translate('admin.userManagement.createrole.createRoleRequest.start'),
+      type: 'loading',
+    });
+  }
+
+  private readonly onSuccesFn = () => {
+    const message = this.transloco.translate(
+      'admin.userManagement.createrole.createRoleRequest.success'
+    );
+
+    this.toastService.open({ type: 'success', message });
+    this.closeModal(true);
+  };
+
+  private readonly onErrorFn = () => {
+    const message = this.transloco.translate(
+      'admin.userManagement.createrole.createRoleRequest.error'
+    );
+
+    this.toastService.open({ message, type: 'danger' });
+  };
+
+  private buildEicFunctionOptions() {
+    this.transloco
```

Comment:
Consideration: Maybe we should create a shared function for this. We are doing the same in wholesale.

Reply:
I've noticed the repetition as well. I was thinking of creating a directive that can be added to the dropdown (e.g. https://timdeschryver.dev/blog/use-angular-directives-to-extend-components-that-you-dont-own#directives-that-load-data). If we go with a function instead, then we'd probably need to return an Observable and subscribe to it in the template. Which is also fine. In the function above, the subscription is in the component.

I'll make the refactor in a separate PR.